File: data-structures/suffix-array/cpp/suffixArray.cpp
#include "SuffixArray.h"
#include <algorithm>

SuffixArray::SuffixArray(const std::string& text) : text(text) {}

void SuffixArray::constructSuffixArray() {
    // Generate suffixes of the text
    int n = text.length();
    std::vector<std::pair<std::string, int>> suffixes;
    for (int i = 0; i < n; i++) {
        suffixes.emplace_back(text.substr(i), i);
    }

    // Sort the suffixes lexicographically
    std::sort(suffixes.begin(), suffixes.end());

    // Populate the suffix array with the sorted suffix indices
    suffixArray.resize(n);
    for (int i = 0; i < n; i++) {
        suffixArray[i] = suffixes[i].second;
    }
}

void SuffixArray::displaySuffixArray() const {
    // Display the suffix array
    for (int i : suffixArray) {
        std::cout << i << " ";
    }
    std::cout << std::endl;
}

File: data-structures/suffix-array/java/SuffixArray.java
import java.util.Arrays;

public class SuffixArray {
    private final String text;
    private final int[] suffixArray;

    public SuffixArray(String text) {
        this.text = text;
        this.suffixArray = new int[text.length()];
    }

    public void constructSuffixArray() {
        Suffix[] suffixes = new Suffix[text.length()];
        for (int i = 0; i < text.length(); i++) {
            suffixes[i] = new Suffix(i, text.substring(i));
        }

        Arrays.sort(suffixes);

        for (int i = 0; i < text.length(); i++) {
            suffixArray[i] = suffixes[i].index;
        }
    }

    public void displaySuffixArray() {
        for (int index : suffixArray) {
            System.out.print(index + " ");
        }
        System.out.println();
    }

    private static class Suffix implements Comparable<Suffix> {
        int index;
        String suffix;

        Suffix(int index, String suffix) {
            this.index = index;
            this.suffix = suffix;
        }

        @Override
        public int compareTo(Suffix other) {
            return this.suffix.compareTo(other.suffix);
        }
    }
}
